synfig-studio/src/gui/duckmatic.cpp

```diff
@@ -851,35 +841,69 @@ Duckmatic::snap_point_to_grid(const synfig::Point& x)const
 	Point ret(x);
 	float radius(0.1/zoom);
 
-	GuideList::const_iterator guide_x,guide_y;
-	bool has_guide_x(false), has_guide_y(false);
-
-	guide_x=find_guide_x(ret,radius);
-	if(guide_x!=guide_list_x_.end())
-		has_guide_x=true;
-
-	guide_y=find_guide_y(ret,radius);
-	if(guide_y!=guide_list_y_.end())
-		has_guide_y=true;
+	GuideList::const_iterator guide;
+	Guide* second_best_guide_match = nullptr;
+	guide = find_guide(ret,radius, second_best_guide_match);
+	bool has_guide = guide != guide_list_.end();
 
 	if(get_grid_snap())
 	{
 		Point snap(
 			floor(ret[0]/get_grid_size()[0]+0.5)*get_grid_size()[0],
 			floor(ret[1]/get_grid_size()[1]+0.5)*get_grid_size()[1]);
-
-		if(std::fabs(snap[0]-ret[0])<=radius && (!has_guide_x || std::fabs(snap[0]-ret[0])<=std::fabs(*guide_x-ret[0])))
-			ret[0]=snap[0],has_guide_x=false;
-		if(std::fabs(snap[1]-ret[1])<=radius && (!has_guide_y || std::fabs(snap[1]-ret[1])<=std::fabs(*guide_y-ret[1])))
-			ret[1]=snap[1],has_guide_y=false;
+		if ( !has_guide || guide->angle == synfig::Angle::deg(90) || guide->angle == synfig::Angle::deg(0)) {
+			if(std::fabs(snap[0]-ret[0])<=radius && (!has_guide || (guide->angle == synfig::Angle::deg(90) &&
+													 std::fabs(snap[0]-ret[0])<=std::fabs((guide->point)[0]-ret[0]))))
+				ret[0]=snap[0],has_guide=false;
+			if(std::fabs(snap[1]-ret[1])<=radius && (!has_guide || (guide->angle == synfig::Angle::deg(0) &&
+													 std::fabs(snap[1]-ret[1])<=std::fabs((guide->point)[1]-ret[1]))))
+				ret[1]=snap[1],has_guide=false;
+		} else {
+			float distance_from_grid = pow( pow( snap[1] - ret[1], 2.0) + pow(snap[0] - ret[0] , 2.0), 0.5);
+			if ( distance_from_grid <= calculate_distance_from_guide(*guide, ret)){
+				ret = snap;
+				has_guide = false;
+			}
+		}
```

Comment:
You over complicated it here. :)
The distance from a point to a grid line is just the difference (subtraction) between one of its coordinates and the grid coordinate.
You would have to compare:
- the distance of the point to the nearest horizontal grid line;
- the distance of the point to the nearest vertical grid line; and
- if guideline snap is enabled, the distance of the point to the nearest guide line

If the distance to the nearest 'snapper' is bigger than the acceptance radius, don't snap.
Otherwise, :
- if the 'snapper' is the grid, force the correspondent value of the grid coordinate
- if the 'snapper' is a guideline, compute the nearest point in the guideline

Reply:
oh ok, I edited it now but I am still not sure if the new edit is exactly what you mean.